MdeModulePkg/Core/Dxe/Mem/Page.c

```diff
@@ -1765,6 +1772,157 @@ CoreFreePages (
+  CoreReleaseMemoryLock ();
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // The free pages code will attempt to zero the freed pages if DebugClearMemory is enabled. However, there is no
+  // guarantee that the memory is writeable, which can cause the core to crash. Furthermore, if DebugClearMemory is
+  // not enabled and NX memory protection is not enabled, we can end up passing pages freed here directly to another
+  // caller in the next allocate pages call, which may not be writeable, which is the expection. So, we need to ensure
+  // the memory is writeable before we attempt to zero it or pass it out. We must do this outside of the memory lock to
+  // avoid deadlocks going through the GCD, which may cause page allocations to occur. It is safe to call the GCD from
+  // here, because the GCD only uses pool backed memory and CoreFreePoolPages() does not call through this path, so we
+  // won't hit a deadlock on the GCD lock. We don't attempt to call this before gCpu is initialized, because before
+  // that point, page attributes can't be set anyway and we may be very early in GCD init and not have the GCD
+  // populated yet.
+  if (gCpu == NULL) {
+    return CoreInternalFreePages (Memory, NumberOfPagesAligned, MemoryType, IsGuarded);
+  }
+
+  CurrentAddress = Memory;
+  EndAddress     = Memory + EFI_PAGES_TO_SIZE (NumberOfPagesAligned);
+  while (CurrentAddress < EndAddress) {
+    Status = CoreGetMemorySpaceDescriptor (
+               CurrentAddress,
+               &Descriptor
+               );
+
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a: failed to get memory space descriptor for %llx with %llx pages\n",
+        __func__,
+        Memory,
+        NumberOfPages
+        ));
+      OperationStatus = Status;
+      ASSERT_EFI_ERROR (Status);
+
+      // if we can't get the descriptor, we don't know what base address to try next, so all we can do is try one
+      // page forward
+      CurrentAddress += EFI_PAGE_SIZE;
+      continue;
+    }
+
+    DescEnd = Descriptor.BaseAddress + Descriptor.Length;
+    // ensure that we only change the attributes for the range that we are interested in, not the entire descriptor,
+    // we may also be in the middle of a descriptor, so ensure our length is not larger than the descriptor length
+    if (EndAddress > DescEnd) {
+      CurrentLength = DescEnd - CurrentAddress;
+    } else {
+      CurrentLength = EndAddress - CurrentAddress;
+    }
+
+    // we always set the attributes here, don't check the descriptor attributes, because they may be out of sync
+    // if the attributes were changed with the CPU arch protocol or the memory attribute protocol
+    Status = CoreSetMemorySpaceAttributes (
+               CurrentAddress,
+               CurrentLength,
+               (Descriptor.Attributes & (~(EFI_MEMORY_RO | EFI_MEMORY_RP)))
+               );
+
+    if (EFI_ERROR (Status)) {
+      DEBUG ((

```

Comment:
Same here.  Is this really an error condition and an ASSERT() condition?  Since it will continue in release builds, should this be a WARN message with no ASSERT() so the log will should some memory pages were not actually freed?

Reply:
This one I had similar logic to making it an error and assert, in that if we are leaking memory, that is an error to me and something I would want to catch with an assert (especially because failing to set attributes may mean something larger with the core is happening that is wrong). But again, in release builds having an error vs a warning doesn’t make a difference (unless the platform is not printing one of the error levels), the log would show leaked pages.

If you prefer to change to warn and no assert, I’ll do that.

Reply:
I dropped this to a WARN, removed the assert, and did not capture any status from this case (i.e. caller would see EFI_SUCCESS if other parts succeed and this fails).